### ISSUE TITLE
systemd wakes up on mouse movement, wasting CPU

### ISSUE THREAD

**Issue by kennylevinsen** [systemd/systemd#14825]
**systemd version the issue has been seen with**

systemd 244 (244.1-1-arch)

**Used distribution**

Arch Linux

**Expected behaviour you didn't see**

init should not wake on mouse input

**Unexpected behaviour you saw**

init consistently uses 0.7-1.3% CPU as during mouse movement when a Wayland compositor is running.

**Steps to reproduce the problem**

1. Use a low-power system, like my i7-8550U (an XPS 13)
2. Run a base Arch install
3. Launch weston or sway
4. Monitor /bin/init in htop
5. Continously wiggle the cursor.

Note that you will also likely see systemd-login waking up, even more than /bin/init. This is due to default `Handle*` configuration of logind, which leads to logind attempting to listen for system buttons on your mouse event device.

Logind event device polling can be disabled by adding the following to logind.conf:
```
HandlePowerKey=ignore
HandleSuspendKey=ignore
HandleHibernateKey=ignore
HandleLidSwitch=ignore
HandleLidSwitchExternalPower=ignore
HandleLidSwitchDocked=ignore
```

**More info**

When I attached a debugger I didn't really see systemd doing anything other than waking up from its poll and going back to sleep *a lot*, suggesting it's polling the mouse event device for some reason.

However, extended debugging leads to a dead init and thus a panic for me, so I didn't get that far with the debugging.

**Comment by poettering** [comment]
pid1 is not opening any input device (except if there's some .socket unit that uses ListenSpecial=/dev/input/event…

what makes you think pid1 wakes from mouse events? there are many reasons it might wakeup. please provide an strace output or so about this...

logind uses EVIOCSMASK to make sure we do not wake up for any input events but the ones we care about. Moreover after opening it checks if the input device actually has a power key or something else we are einterested in with the EVIOCGBIT ioctl, and closes all devices that do not immediately. Finally, we do not even go as far as opening any devices that might not be suitable, as we only open those that carry a udev tag of "power-switch", and that's only those matching this:

```SUBSYSTEM=="input", KERNEL=="event*", ENV{ID_INPUT_SWITCH}=="1", TAG+="power-switch"
SUBSYSTEM=="input", KERNEL=="event*", ENV{ID_INPUT_KEY}=="1", TAG+="power-switch"
```

htop is not a tracing tool, it's not suitable for tracking this down. A simple first step would be strace logind and pid1, and provide relevant output here.

both pid1 and logind provide IPC interfaces, so apps can call them any time they like? maybe your desktop environment does that needlessly? maybe it calls logind's SetIdleHint() a bit too frequently, and keeps flooding logind with that? there's nothing we can do about that...

Anyway, please provide some better data. Otherwise I am inclined to just close this, as there's no indication that logind/pid1 really do anything wrong as the used tools are not suitable for determining that.

**Comment by kennylevinsen** [comment]
Yes, when I was debugging this, I did see that logind tries to avoid irrelevant devices.

I didn't get much further in my debugging after discovering the logind workaround, as I don't have a good systemd development environment: debugging logind crashes my compositor after a while, debugging systemd panics my kernel after while.

Data incoming as requested:

### `perf record`

I have done several runs of `perf record` to gather data for the following:

- systemd and systemd-logind `cycle` reports
- With default Handle actions, and with all Handle actions set to "Ignore"
- With and without cursor jiggling

Notes about test environment:
- Pretty empty Arch install, everything non-essential stopped and/or killed (process lists included in test results)
- No desktop environment, only a bare Wayland compositor rendering nothing but a cursor to a hardware plane.
- Compositor only implements logind seat/drm management, nothing more (e.g. TakeDevice/ReleaseDevice/... - specifically no SetIdleHint support).
- `busctl monitor`/`busctl --user monitor` are silent during test.
- Tests are run timed by a script from a different VT (scripts included in test results), cursor wiggled kept as consistent as I could.

The results from my last run are as follows:

| Binary  | Handle action state | Events without jiggle | Events with jiggle |
|---------|---------------------|-----------------------|--------------------|
| logind  | default             | 0                     | 550                |
| logind  | ignore              | 0                     | 0                  |
| systemd | default             | 0                     | 814                |
| systemd | ignore              | 15                    | 356                |

I ran these tests multiple times, and saw no major deviations (the exact event counts would differ, and sometimes no-jiggle runs would get a dozen events, but nothing affecting the conclusion). `perf record` results match entirely as expected the results of process monitoring tools like `htop`.

The last results, including the scripts, logind config and process list for each run can be seen here: [perf.zip](https://github.com/systemd/systemd/files/4422006/perf.zip). Note that this is done on the systemd

Looking at the traces themselves, both systemd and logind appears to primarly be assigned blame for kernel work, likely as a function of our poll. Note that these are traces are taken on my regular setup with a release build of v245 (245.3-2-arch), so it's a bit low on debug info.

### `strace` 

Strace shows us as hanging in the epoll_wait. This seems in line with `perf` above considering the work to be kernel-side as a result of our poll request.

---

Tell me if there's anything else that could be useful here.

**Comment by poettering** [comment]
hmm, so if you say we just hang in epoll_wait() while you wiggle then that means logind/pid1 are really not doing anything, so we can close this, no?

i.e. there's nothing we actually do wrong?

**Comment by kennylevinsen** [comment]
Telling the kernel to do useless work is not "nothing wrong". The question is what exactly is causing it, and whether it can be avoided.

**Comment by kennylevinsen** [comment]
So, digging into things a bit:

1. PID1's cycles are from the fd store. logind sends an fd to all its session devices, and pid1 makes a blank poll on them to wait for errors. For an active evdev device (like a mouse) on a weak machine (e.g. my laptop running at 800MHz), the work of servicing the evdev_client and the poll filtering uninteresting waitqueue wakeups is apparently enough to be noticable.

2. logind's cycles are from the button poll. This evdev client does have a filter applied, but we can be woken up regardless due to each evdev instance sharing a waitqueue across all its clients.

3. logind is listening on my mouse due to logitech_hidpp misreporting Unifying wireless mice as having keyboard capabilities with KEY_SLEEP and KEY_POWER, so it slips through logind's filtering and gets actively polled.

Looking at the evdev code, I do suspect that the penalty of the empty poll could be minimized a bit, which I might take a look at. It will probably never be free for PID1 to poll those devices, though.

**Comment by poettering** [comment]
hmm, can we close this? @kennylevinsen elsewhere you mentioned you had success on the kernel side? did anything get merged? have a link?

**Comment by kennylevinsen** [comment]
Yes, I was waiting for all the things to come into place so that I could have a nice "everything is fixed" closure. :)

Reduction of unwanted wakeups when polling only for EPOLLERR|EPOLLHUP by using keyed wakeups:
- https://marc.info/?l=dri-devel&m=158677018523208&w=2 (drm)
- https://marc.info/?l=linux-input&m=158656213030584&w=2 (evdev)

These two affect PID1, and are queued for 5.8

Reduction of unwanted wakeups in evdev when multiple clients are polling with different filtering settings by using per-client waitgroups:
- https://marc.info/?l=linux-input&m=158818570014786&w=2

This one is pending review, and affects logind's polling of system buttons, which is especially bad when mice report presence of system buttons (like Logitech Unifi mice do).

**Comment by poettering** [comment]
thanks so much for tracking this down so thoroughly. great work!

**Comment by kennylevinsen** [comment]
The PID1 fixes have been released as part of 5.8. :)
